src/index.ts

```diff
@@ -137,15 +137,20 @@ type CasesExhaustive<A extends AnyMember, B> = {
  * Enables a {@link match} expression to cover only some cases provided a
  * wildcard case is declared with which to match the remaining cases.
  */
-type CasesWildcard<A extends AnyMember, B> = Partial<CasesExhaustive<A, B>> & {
+// Don't reuse `CasesExhaustive` here (via `Partial`), it causes exhaustive
+// errors to always point to the absence of a wildcard.
+type CasesWildcard<A extends AnyMember, B> = {
+  readonly [V in A as Tag<V>]?: (val: Value<V>) => B
+} & {
```

Comment:
I wonder if we could still re-use `CasesExhaustive` in the mapped type, just so we don't have to duplicate the definition of `CasesExhaustive`.

```suggestion
type CasesWildcard<A extends AnyMember, B> = {
  [K in keyof CasesExhaustive<A, B>]?: CasesExhaustive<A, B>[K]
} & {
```

Reduced test case which seems to work:

```ts
type F<A extends string> = H<A> | G<A>;
type G<A extends string> = { [K in A]: any };
type H<A extends string> = { [K in keyof G<A>]?: G<A>[K] } & { bad: any };
const x: F<"good"> = {};
```